src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp

```diff
@@ -104,321 +150,710 @@ enum Spells
+
+            while (uint32 eventId = events.ExecuteEvent())
             {
-                DoCastAOE(SPELL_RADIANCE);
+                switch (eventId)
+                {
+                    case EVENT_HOLY_SMITE_E:
+                        if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 50.0f, true))
+                            DoCast(target, SPELL_SMITE);
+                        events.Repeat(2s, 3s);
+                        break;
+                    case EVENT_HOLY_FIRE:
+                        if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 30.0f, true))
+                            DoCast(target, SPELL_HOLY_FIRE);
+                        events.Repeat(9s, 12s);
+                        break;
+                    case EVENT_RENEW:
+                        me->InterruptNonMeleeSpells(true);
+                        if (roll_chance_i(50) && me->GetHealthPct() < 100)
+                            DoCastSelf(SPELL_RENEW);
+                        else
+                        {
+                            if (_memory && _memory->GetHealth() > 1)
+                                DoCast(_memory, SPELL_RENEW);
+                            else
+                                DoCastSelf(SPELL_RENEW);
+                        }
+
+                        events.Repeat(15s, 17s);
+                        break;
+                    case EVENT_SUMMON_MEMORY:
+                        // Memory spawns at random player's position
+                        if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 50.0f, true))
+                            DoCast(target, SPELL_SUMMON_MEMORY, true);
+                        break;
+                    case EVENT_MEMORY_AGGRESSIVE:
+                        if(_memory) {
+                            _memory->SetReactState(REACT_AGGRESSIVE);
+                            _memory->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE);
+                        }
+                        break;
+                    case EVENT_CHALLENGE_DONE:
+                        me->RemoveAllAuras();
+                        me->GetThreatManager().ClearAllThreat();
+                        me->SetFullHealth();
+                        DoCastAOE(SPELL_PALETRESS_ACH, true);
+                        me->RestoreFaction();
+
+                        if (IsHeroic())
+                            DoCastAOE(SPELL_PALETRESS_CONFESSOR, true);
+
+                        _JustDied();
+                        me->SetWalk(true);
+                        me->GetMotionMaster()->MovePoint(POINT_DESPAWN, bossExitPos);
+                        break;
+                    default:
+                        break;
+                }
+            }
 
-                uiRadianceTimer = 16000;
-            } else uiRadianceTimer -= uiDiff;
+            if (!UpdateVictim())
+                return;
 
             DoMeleeAttackIfReady();
         }
+
+    private:
+        bool _memorySummoned;
+        Creature* _memory;
```

Comment:
WorldObject* (and child classes) pointers are not allowed in script classes, they cause a free'd memory access crash if the object is deleted (as easy as a GM deleting it)